### ISSUE TITLE
state['status'] is None exception (virsh.dump testcases)

### ISSUE THREAD

**Issue by svirt** [avocado-framework/avocado#898]
Hello,
I encountered a strange error case which is not repeating exactly during the same step but it occurs during one of the virsh.dump subtests (randomly between  tests # 10-20 of 54 . The exception occurs in avocado/core/result.py

below is the sample ( ./scripts/avocado run virsh.dump --vt-type libvirt ):

 (14/54) type_specific.io-github-autotest-libvirt.virsh.dump.positive_test.non_acl.memory_dump.kdump-snappy_format:  
Avocado crashed: KeyError: None
Traceback (most recent call last):

  File "/home/avocado-jenkins/avocado/avocado/core/job.py", line 492, in _run
    timeout=self.timeout)
  File "/home/avocado-jenkins/avocado/avocado/core/runner.py", line 290, in run_suite
    deadline)
  File "/home/avocado-jenkins/avocado/avocado/core/runner.py", line 246, in run_test
    self.result.check_test(test_state)
  File "/home/avocado-jenkins/avocado/avocado/core/result.py", line 98, in check_test
    output_plugin.check_test(state)
  File "/home/avocado-jenkins/avocado/avocado/core/result.py", line 252, in check_test
    add = status_map[state['status']]

KeyError: None

Please include the traceback info and command line used on your bug report
Report bugs visiting https://github.com/avocado-framework/avocado/issues/new 

Somehow state['status'] has a non-supported None value, as well as e.g. start['time_start'] and start['time_start'] are None which doesn't look good. 
I also tried to increase the timeout value in the virsh.dump test cfg to 60 (because for some of its subtests the default 5 seconds were not enough).

As it's not exactly reproducible on the exact step it makes the debugging quite time consuming (when I launch a single subtest separately it doesn't crash) because of running each time the whole set of subtests.

Maybe there are some ideas/hints about it ?

Thank you in advance

**Comment by ldoktor** [comment]
Thanks for the report, I smell some bug in the runner... I'll look at this tomorrow.

**Comment by svirt** [comment]
**@ldoktor**
Thanks a lot

**Comment by ldoktor** [comment]
Hello @svirt, I was not really able to reproduce it, but I found a possible cause. Please try the https://github.com/avocado-framework/avocado/issues/902 if that fixes your issue.

There are few things I noticed while executing `virsh.dump` tests. There were lots of failures (probably caused by my setup, because I don't use libvirt)

```
 (1/54) type_specific.io-github-autotest-libvirt.virsh.dump.positive_test.non_acl.no_option: PASS (5.00 s)
 (2/54) type_specific.io-github-autotest-libvirt.virsh.dump.positive_test.non_acl.pause_dump.live: PASS (5.16 s)
 (3/54) type_specific.io-github-autotest-libvirt.virsh.dump.positive_test.non_acl.pause_dump.crash: ERROR (6.19 s)
 (4/54) type_specific.io-github-autotest-libvirt.virsh.dump.positive_test.non_acl.pause_dump.reset: ERROR (6.28 s)
 (5/54) type_specific.io-github-autotest-libvirt.virsh.dump.positive_test.non_acl.live_dump: ERROR (1.64 s)
 (6/54) type_specific.io-github-autotest-libvirt.virsh.dump.positive_test.non_acl.crash_dump: ERROR (4.45 s)
 (7/54) type_specific.io-github-autotest-libvirt.virsh.dump.positive_test.non_acl.reset_dump: ERROR (5.38 s)
 (8/54) type_specific.io-github-autotest-libvirt.virsh.dump.positive_test.non_acl.bypass_cache_dump: ERROR (1.86 s)
 (9/54) type_specific.io-github-autotest-libvirt.virsh.dump.positive_test.non_acl.bypass_cache_reset_dump: FAIL (3.38 s)
 (10/54) type_specific.io-github-autotest-libvirt.virsh.dump.positive_test.non_acl.memory_dump.default_format: ERROR (1.46 s)
 (11/54) type_specific.io-github-autotest-libvirt.virsh.dump.positive_test.non_acl.memory_dump.elf_format: ERROR (4.25 s)
 (12/54) type_specific.io-github-autotest-libvirt.virsh.dump.positive_test.non_acl.memory_dump.kdump-zlib_format: ERROR (6.19 s)
 (13/54) type_specific.io-github-autotest-libvirt.virsh.dump.positive_test.non_acl.memory_dump.kdump-lzo_format: ERROR (6.34 s)
 (14/54) type_specific.io-github-autotest-libvirt.virsh.dump.positive_test.non_acl.memory_dump.kdump-snappy_format: SKIP
 (15/54) type_specific.io-github-autotest-libvirt.virsh.dump.positive_test.non_acl.memory_crash_dump: PASS (4.64 s)
 (16/54) type_specific.io-github-autotest-libvirt.virsh.dump.positive_test.non_acl.memory_bypass_cache_dump: ERROR (2.14 s)
 (17/54) type_specific.io-github-autotest-libvirt.virsh.dump.positive_test.non_acl.lzop_format_dump: ERROR (1.50 s)
 (18/54) type_specific.io-github-autotest-libvirt.virsh.dump.positive_test.non_acl.gzip_format_dump: ERROR (6.19 s)
 (19/54) type_specific.io-github-autotest-libvirt.virsh.dump.positive_test.non_acl.bzip2_format_dump: ERROR (1.68 s)
...
```

and a lot of noise in the standard output:

```
 (24/54) type_specific.io-github-autotest-libvirt.virsh.dump.positive_test.acl_test.no_option: \PolicyKit daemon disconnected from the bus.
We are no longer a registered authentication agent.
ERROR (6.19 s)
 (25/54) type_specific.io-github-autotest-libvirt.virsh.dump.positive_test.acl_test.pause_dump.live: -PolicyKit daemon disconnected from the bus.
We are no longer a registered authentication agent.
PolicyKit daemon disconnected from the bus.
We are no longer a registered authentication agent.

** (pkttyagent:11968): WARNING **: Unable to register authentication agent: GDBus.Error:org.freedesktop.DBus.Error.NoReply: Message recipient disconnected from message bus without replying
Error registering authentication agent: GDBus.Error:org.freedesktop.DBus.Error.NoReply: Message recipient disconnected from message bus without replying (g-dbus-error-quark, 4)
PolicyKit daemon reconnected to bus.
Attempting to re-register as an authentication agent.
We are now a registered authentication agent.
ERROR (6.19 s)
 (26/54) type_specific.io-github-autotest-libvirt.virsh.dump.positive_test.acl_test.pause_dump.crash: |PolicyKit daemon disconnected from the bus.
PolicyKit daemon disconnected from the bus.
PolicyKit daemon disconnected from the bus.
We are no longer a registered authentication agent.
We are no longer a registered authentication agent.
We are no longer a registered authentication agent.
PolicyKit daemon disconnected from the bus.
We are no longer a registered authentication agent.
PolicyKit daemon disconnected from the bus.
We are no longer a registered authentication agent.
PolicyKit daemon reconnected to bus.
Attempting to re-register as an authentication agent.
PolicyKit daemon reconnected to bus.
PolicyKit daemon reconnected to bus.
Attempting to re-register as an authentication agent.
Attempting to re-register as an authentication agent.
PolicyKit daemon reconnected to bus.
Attempting to re-register as an authentication agent.
We are now a registered authentication agent.
We are now a registered authentication agent.
We are now a registered authentication agent.
We are now a registered authentication agent.
|^C^C^C^C^C^C^C^ERROR (6.19 s)
^C (27/54) type_specific.io-github-autotest-libvirt.virsh.dump.positive_test.acl_test.pause_dump.reset:  ^C^C^/^C^C^C^C^C^C^C^C^C^C^C^C^-^C^C^C^C^C^C^C^C^C^C^C^C^C^CPolicyKit daemon disconnected from the bus.
We are no longer a registered authentication agent.
PolicyKit daemon disconnected from the bus.
We are no longer a registered authentication agent.
```

and last, I was not able to interrupt the testing (as can be seen from the many `^C`s (@lmr - no need to respond/fix, I'm just letting you know) in the previous output. I had to kill the right pid and there were several to choose from:

```
ps ax |grep virsh
11239 pts/28   Sl+    0:00 python ./scripts/avocado run --vt-type libvirt virsh.dump
11770 pts/28   Sl+    0:00 python ./scripts/avocado run --vt-type libvirt virsh.dump
12478 pts/28   Sl+    0:00 python ./scripts/avocado run --vt-type libvirt virsh.dump
13462 pts/28   Sl+    0:00 python ./scripts/avocado run --vt-type libvirt virsh.dump
14206 pts/28   Sl+    0:00 python ./scripts/avocado run --vt-type libvirt virsh.dump
15258 pts/28   Sl+    0:00 python ./scripts/avocado run --vt-type libvirt virsh.dump
16058 pts/28   Sl+    0:00 python ./scripts/avocado run --vt-type libvirt virsh.dump
16403 pts/28   Sl+    0:00 python ./scripts/avocado run --vt-type libvirt virsh.dump
16590 pts/28   Dl+    0:00 python ./scripts/avocado run --vt-type libvirt virsh.dump
17637 pts/28   Dl+    0:00 python ./scripts/avocado run --vt-type libvirt virsh.dump
17683 pts/29   S+     0:00 grep --color=auto virsh
31029 pts/28   S+     0:00 sudo PYTHONPATH=.:.:.:.::/home/medic/.eclipse/org.eclipse.platform_4.5.0_1473617060_linux_gtk_x86_64/plugins/org.python.pydev_4.2.0.201507041133/pysrc:/home/medic/.eclipse/org.eclipse.platform_4.5.0_1473617060_linux_gtk_x86_64/plugins/org.python.pydev_4.2.0.201507041133/pysrc:/home/medic/.eclipse/org.eclipse.platform_4.5.0_1473617060_linux_gtk_x86_64/plugins/org.python.pydev_4.2.0.201507041133/pysrc ./scripts/avocado run --vt-type libvirt virsh.dump
31030 pts/28   S+     0:04 python ./scripts/avocado run --vt-type libvirt virsh.dump
```

@svirt do you also have those issues, or is it just my setup?

**Comment by svirt** [comment]
**@ldoktor**
Thanks a lot, I will check this fix. I guess I didn't notice similar issues in the log, but maybe I am wrong as I was more carefully analyzing the traceback related parts of the log

**Comment by svirt** [comment]
**@ldoktor** 
I checked and yes I see "PolicyKit daemon disconnected from the bus." errors too..

**Comment by ldoktor** [comment]
And how about the `state['status'] is None exception`s, did the patch help?

**Comment by svirt** [comment]
**@ldoktor**
I didn't finish checking it yesterday,  seems some other files from your repo-fork should be applied too for the patch to work, so I will continue with it

**Comment by svirt** [comment]
**@ldoktor**
I mean after your patch appeared *AttributeError: 'Job' object has no attribute 'funcatexit'* so i am trying to fix it locally here in order to perform the test

**Comment by svirt** [comment]
**@ldoktor**
So I took your runner.py, job.py and data_structures.py, ran virsh.dump tests completely and it worked without an avocado crash! (the failed tests I will debug deeper  later on). Will you merge those needed patches?

**Comment by ldoktor** [comment]
Oh, `funcatexit` was added in this (or last) version. Yes, it impacted avocado-vt and the runner is quite extensively rewritten and polished. Note for me, always ask about the version the bug occurs on :smile:

**Comment by svirt** [comment]
**@ldoktor**
Thank you for your fix !

**Comment by ldoktor** [comment]
You're welcome. We need to decide on one little nitpick from Cleber and then it'll be included in master (probably today). If you use RPM then you have to wait for a few weeks (or build one yourself)

**Comment by svirt** [comment]
**@ldoktor**
np, its ok, i dont use the git version

**Comment by svirt** [comment]
**@ldoktor** 
Hi , I am back again to this issue :) I fully updated last week avocado and avocado-vt from the git to have all the latest patches you mentioned above (before I had some problems with applying different single patches as there were dependencies I guess). The virsh.dump test still fails the same way (I will check also if the test itself is modified recently in git or not)

**Comment by ldoktor** [comment]
Hi @svirt, thank you for the report. I'm finally able to reproduce the problem, it's related to sending the state of a failed test. I'll work on this today/tomorrow (lot's of meetings today).

**Comment by svirt** [comment]
**@ldoktor** 
Thanks a lot. Probably there is also something related to timeouts (my assumption) , because I saw in some of the logs timeout related messages

**Comment by ldoktor** [comment]
Btw the latest commit from Lucas https://github.com/avocado-framework/avocado/pull/954 should at least minimize the damage. Anyway it's not preventing the failure...

**Comment by svirt** [comment]
I am pasting some recent logs related to timeout (if this can help you)

07:20:50 env_process      L0869 ERROR| Timeout reached waiting for type_specific.io-github-autotest-libvirt.virsh.dump.positive_test.non_acl.pause_dump.live to end
07:20:50 process          L0275 INFO | Running 'true'
07:20:50 process          L0275 INFO | Running 'ps -o comm 1'
07:20:50 process          L0362 DEBUG| [stdout] COMMAND
07:20:50 process          L0362 DEBUG| [stdout] systemd
07:20:50 stacktrace       L0036 ERROR|
07:20:50 stacktrace       L0039 ERROR| Reproduced traceback from: /usr/lib/python2.7/site-packages/avocado_plugins_vt-0.31.0-py2.7.egg/avocado_vt/test.py:382
07:20:50 stacktrace       L0042 ERROR| Traceback (most recent call last):
07:20:50 stacktrace       L0042 ERROR|   File "/usr/lib/python2.7/site-packages/avocado_plugins_vt-0.31.0-py2.7.egg/avocado_vt/test.py", line 158, in runTest
07:20:50 stacktrace       L0042 ERROR|     self._runTest()
07:20:50 stacktrace       L0042 ERROR|   File "/usr/lib/python2.7/site-packages/avocado_plugins_vt-0.31.0-py2.7.egg/avocado_vt/test.py", line 326, in _runTest
07:20:50 stacktrace       L0042 ERROR|     env_process.postprocess(self, params, env)
07:20:50 stacktrace       L0042 ERROR|   File "/usr/lib/python2.7/site-packages/avocado_plugins_vt-0.31.0-py2.7.egg/virttest/error_context.py", line 135, in new_fn
07:20:50 stacktrace       L0042 ERROR|     return fn(*args, **kwargs)
07:20:50 stacktrace       L0042 ERROR|   File "/usr/lib/python2.7/site-packages/avocado_plugins_vt-0.31.0-py2.7.egg/virttest/env_process.py", line 1141, in postprocess
07:20:50 stacktrace       L0042 ERROR|     raise virt_vm.VMError("Failures occurred while postprocess:%s" % err)
07:20:50 stacktrace       L0042 ERROR| VMError: Failures occurred while postprocess:
07:20:50 stacktrace       L0042 ERROR| Postprocess: Timeout reached waiting for type_specific.io-github-autotest-libvirt.virsh.dump.positive_test.non_acl.pause_dump.live to end
07:20:50 stacktrace       L0043 ERROR|
07:20:50 test             L0485 ERROR| Traceback (most recent call last):

07:20:50 test             L0485 ERROR|   File "/usr/lib/python2.7/site-packages/avocado_plugins_vt-0.31.0-py2.7.egg/avocado_vt/test.py", line 424, in _run_avocado
    raise test_exception

07:20:50 test             L0485 ERROR| VMError: Failures occurred while postprocess:
Postprocess: Timeout reached waiting for type_specific.io-github-autotest-libvirt.virsh.dump.positive_test.non_acl.pause_dump.live to end

07:20:50 test             L0502 ERROR| ERROR type_specific.io-github-autotest-libvirt.virsh.dump.positive_test.non_acl.pause_dump.live -> VMError: Failures occurred while postprocess:
Postprocess: Timeout reached waiting for type_specific.io-github-autotest-libvirt.virsh.dump.positive_test.non_acl.pause_dump.live to end

**Comment by svirt** [comment]
Recently I also encountered the following type of errors (haven't fully debugged yet)

08:26:56 stacktrace       L0039 ERROR| Reproduced traceback from: /usr/lib/python2.7/site-packages/avocado_plugins_vt-0.31.0-py2.7.egg/avocado_vt/test.py:382
08:26:56 stacktrace       L0042 ERROR| Traceback (most recent call last):
08:26:56 stacktrace       L0042 ERROR|   File "/usr/lib/python2.7/site-packages/avocado_plugins_vt-0.31.0-py2.7.egg/avocado_vt/test.py", line 158, in runTest
08:26:56 stacktrace       L0042 ERROR|     self._runTest()
08:26:56 stacktrace       L0042 ERROR|   File "/usr/lib/python2.7/site-packages/avocado_plugins_vt-0.31.0-py2.7.egg/avocado_vt/test.py", line 333, in _runTest
08:26:56 stacktrace       L0042 ERROR|     env.save()
08:26:56 stacktrace       L0042 ERROR|   File "/usr/lib/python2.7/site-packages/avocado_plugins_vt-0.31.0-py2.7.egg/virttest/utils_env.py", line 187, in save
08:26:56 stacktrace       L0042 ERROR|     f = open(filename, "w")
08:26:56 stacktrace       L0042 ERROR| IOError: [Errno 2] No such file or directory: '/var/tmp/avocado_jHQNXF/env'
08:26:56 stacktrace       L0043 ERROR|
08:26:56 test             L0485 ERROR| Traceback (most recent call last):

08:26:56 test             L0485 ERROR|   File "/usr/lib/python2.7/site-packages/avocado_plugins_vt-0.31.0-py2.7.egg/avocado_vt/test.py", line 424, in _run_avocado
    raise test_exception

08:26:56 test             L0485 ERROR| IOError: [Errno 2] No such file or directory: '/var/tmp/avocado_jHQNXF/env'

08:26:56 test             L0502 ERROR| ERROR type_specific.io-github-autotest-libvirt.virsh.dump.positive_test.non_acl.memory_dump.default_format -> IOError: [Errno 2] No such file or directory: '/var/tmp/avocado_jHQNXF/env'

OR

Exception in thread tail_thread_pjUosgmQ_virsh -c q:
Traceback (most recent call last):
  File "/usr/lib64/python2.7/threading.py", line 811, in __bootstrap_inner
    self.run()
  File "/usr/lib64/python2.7/threading.py", line 764, in run
    self.__target(*self.__args, **self.__kwargs)
  File "/usr/lib/python2.7/site-packages/aexpect/client.py", line 585, in _tail
    status = self.get_status()
  File "/usr/lib/python2.7/site-packages/aexpect/client.py", line 278, in get_status
    wait_for_lock(self.lock_server_running_filename)
  File "/usr/lib/python2.7/site-packages/aexpect/shared.py", line 37, in wait_for_lock
    fd = get_lock_fd(filename)
  File "/usr/lib/python2.7/site-packages/aexpect/shared.py", line 10, in get_lock_fd
    open(filename, "w").close()
IOError: [Errno 2] No such file or directory: '/tmp/aexpect_pjUosgmQ/lock-server-running'

**Comment by ldoktor** [comment]
OK, not that simple as I thought :smile:. I think the original problem disappeared with the Lucas' patch mentioned earlier. Anyway your new reports are much more interesting. The main problem here is, that `io-github-autotest-libvirt` forks itself and uses SIGUSR1 to kill the `check_bypass`.

This is wrong, because `io-github-autotest-libvirt` never set the `SIGUSR1` signal handler. It only hopes for default handler (kill) kicks-in. As avocado sets the handler, it raises exception and proceeds. And this is where it gets nasty, because the forked process does not die, it catches the exception and proceeds with cleanup. But at the same time the original process also proceeds to cleanup so the second one fails to cleanup the environment, because the files are already removed.

So unless I'm missed something (which is possible, I saw the `/usr/share/avocado/data/avocado-vt/test-providers.d/downloads/io-github-autotest-libvirt/libvirt/tests/src/virsh_cmd/domain/virsh_dump.py` file for the first time) there are at least two bugs in it and the signal handler just made them visible. I'll send PR in `io-github-autotest-libvirt` and we'll see, maybe you or people there will explain why am I wrong.

On the other hand we should revisit the idea of custom `SIGUSR1` (and `SIGINT`) handlers as people might use it for other purposes (although `SIGTERM` should be more appropriate when they wish the default handler). Note that ideally it should not prevent people from using them as far as they set their handlers. If they hope for default handlers than we break their testing (as happened here).

Another problem on Avocado side can be, that we only wait for the original spawned PID. If one creates a test, which spawns several workers and reports the results from them, then avocado might not behave correctly (it can finish early).

Last thing quite important is, that the test postprocess is part of the test itself (Including stdout/stderr checker and all those). I remember I wanted to remove them, because in my opinion all those should live in `runner` instead. And this is another example why. Because if the avocado test is forked, the whole postprocess happens twice. This would be avoided if the Test itself would be only a pure test and the `runner` would call and catch for exceptions. _(ps this will be tricky, because avocado-vt uses custom `run_avocado` function. This only works because the `run_avocado` is part of the test and would not work otherwise. So we'll have to change quite some code...)_

So to wrap it up:

1. I'll send a PR in `io-github-autotest-libvirt`, which should fix the issues you're seeing nowadays
2. We need to think whether setting signal handlers is the right way
3. We should reconsider the `run_avocado` method, which is prone to failures when users use fork inside test

@clebergnu I'm wondering if you can find some holes in my thoughts...

**Comment by svirt** [comment]
**@ldoktor**
Thanks a lot for your detailed explanation. Seems like the problem is hidden quite deep inside and is affected by many conditions. Maybe that's why it happens only with this virsh.dump test (as far as I checked so far it's like that, at least the the last time). Probably here is the case where all the unwanted conditions are met and that's why we have permanent errors.

**Comment by ldoktor** [comment]
Hi @svirt, thanks for sharing your time with me, I'm glad to report that I tested everything. There were several causes, fixes are here:

https://github.com/avocado-framework/avocado/pull/990 - which sets the timeout properly from "avocado_params" instead of "cartesian params".
https://github.com/avocado-framework/avocado-vt/pull/353 - which is avocado-vt part of the params handling patch
https://github.com/autotest/tp-libvirt/pull/693 - which uses SIGTERM instead of SIGUSR1 which has handler set by avocado and causes double cleanup because of forked process. The options are either to set SIGUSR1 handler or to use SIGTERM which does exactly to same as default SIGUSR1 handler.

The reason I was unable to reproduce the @svirt timeout problem was, that Beijing guys already fixed the `timeout` problem by using different name for the variable in cartesian, so it worked fine on the latest libvirt tests, but @svirt used the older version which did not contain the fix.

@svirt please test those paches on your machine, I tested it on F23 and all tests passed (the `avocado.core.runner` workaround is not needed anymore)

**Comment by svirt** [comment]
**@ldoktor**
Hi, thank you very much for your help and support!
Yesterday at the end of the work day I spent testing the virsh tests with the modifications applied (plus timeout one) and everything worked fine! All the tests PASSed.  It was really due to the different name of the timeout variable

**Comment by svirt** [comment]
**@ldoktor**
I will test those new patches and will let you

**Comment by svirt** [comment]
**@ldoktor**
I applied and checked your params patches. The virsh.dump tests PASSed, the only thing I had to modify a little was to locally revert back the TestNAError (seems it was replaced by TestSkipError) as ..kdump-lzo_format and ..kdump-snappy_format tests were giving out an ERROR (instead of SKIP). I guess this doesn't affect the params patch behavior anyway. Otherwise, except this, all looks good.
